library/pk_internal.h

```diff
@@ -49,11 +49,19 @@ int mbedtls_pk_update_public_key_from_keypair(mbedtls_pk_context *pk);
 
 #if defined(MBEDTLS_ECP_LIGHT) && !defined(MBEDTLS_ECP_C)
 /**
- * This is the same function defined in pk.h, but with different guards. The
- * purpose is to make this function private as long as ECP_C is not defined,
- * but ECP_LIGHT is defined.
+ * This is the same function defined in pk.h, but with the following
+ * differences:
+ * - it has different guards
+ *     - the purpose is to eliminate this function in the future, so for now it
+ *       is just set as private as soon as ECP_C is not defined because there is
+ *       some library's code that's using it. However in the future also this
+ *       internal version will be removed.
+ * - it returns a const pointer
+ *     - because we don't want any caller to have an explicit write persmission
+ *       on the pk_ctx data. This would make upcoming PSA focused improvements
+ *       more difficult to achieve.
  */
-static inline mbedtls_ecp_keypair *mbedtls_pk_ec(const mbedtls_pk_context pk)
+static inline const mbedtls_ecp_keypair *mbedtls_pk_ec(const mbedtls_pk_context pk)
```

Comment:
I don't like having two function with the same name but different signatures and contracts. I'd prefer having three functions for clarity:
1. Public `mbedtls_pk_ec()` - formally deprecated, hence never used by the library or tests (it is absent when `MBEDTLS_DEPRECATED_REMOVED` is set).
2. Internal function `mbedtls_pk_ec_ro()` returning a `const` pointer.
3. Internal function `mbedtls_pk_ec_rw()` returning a non-`const` pointer.

Reply:
Should have updated my comment, because we followed exactly the same path in private discussion with Valerio:
- it's to early to formally deprecate `pk_ec`, we must clarify replacements first;
- this should be its own PR indeed.